Mathlib/Computability/AkraBazzi/AkraBazzi.lean

```diff
@@ -0,0 +1,1505 @@
+                    _ = (fun (n:ℕ) => ((1:ℝ) ^ 2))        := by simp
+                    _ =o[atTop] (fun (n:ℕ) => (log n)^2)  := by
+                          refine IsLittleO.rpow (by linarith) ?nonneg ?log
+                          case nonneg =>
+                            filter_upwards [eventually_ge_atTop 1] with n hn
+                            have hn' : (1:ℝ) ≤ (n:ℝ) := by
+                              convert_to ((1:ℕ) : ℝ) ≤ (n:ℝ)
+                              · exact Nat.cast_one.symm
+                              · exact Nat.cast_le_cast hn
+                            simp [log_nonneg hn']
+                          case log => exact IsLittleO.nat_cast_atTop (f := fun _ => 1)
+                                        <| isLittleO_const_log_atTop
+         _ = (fun (n:ℕ) => (n:ℝ))  := by ext; simp
+
+lemma eventually_b_le_r : ∀ᶠ (n:ℕ) in atTop, ∀ i, (b i : ℝ) * n - (n / log n ^ 2) ≤ r i n := by
+  filter_upwards [R.dist_r_b'] with n hn
+  intro i
+  have h₁ : 0 ≤ b i := le_of_lt <| R.b_pos _
+  rw [sub_le_iff_le_add, add_comm, ←sub_le_iff_le_add]
+  calc (b i : ℝ) * n - r i n = ‖b i * n‖ - ‖(r i n : ℝ)‖ := by
+                            simp only [norm_mul, IsROrC.norm_natCast, sub_left_inj,
+                                       Nat.cast_eq_zero, Real.norm_of_nonneg h₁]
+                         _ ≤ ‖(b i * n : ℝ) - r i n‖     := norm_sub_norm_le _ _
+                         _ = ‖(r i n : ℝ) - b i * n‖      := norm_sub_rev _ _
+                         _ ≤ n / log n ^ 2                := hn i
+
+lemma eventually_r_le_b : ∀ᶠ (n:ℕ) in atTop, ∀ i, r i n ≤ (b i : ℝ) * n + (n / log n ^ 2) := by
+  filter_upwards [R.dist_r_b'] with n hn
+  intro i
+  calc r i n = b i * n + (r i n - b i * n)   := by ring
+             _ ≤ b i * n + ‖r i n - b i * n‖   := by gcongr; exact Real.le_norm_self _
+             _ ≤ b i * n + n / log n ^ 2                   := by gcongr; exact hn i
+
+lemma eventually_r_lt_n : ∀ᶠ (n:ℕ) in atTop, ∀ i, r i n < n := by
+  filter_upwards [eventually_ge_atTop R.n₀] with n hn
+  exact fun i => R.r_lt_n i n hn
+
+lemma eventually_bi_mul_le_r : ∀ᶠ (n:ℕ) in atTop, ∀ i, (b R.min_bi / 2) * n ≤ r i n := by
+  have gt_zero : 0 < b R.min_bi := R.b_pos R.min_bi
+  have hlo := isLittleO_self_div_log_id
+  rw [Asymptotics.isLittleO_iff] at hlo
+  have hlo' := hlo (by positivity : 0 < b R.min_bi / 2)
+  filter_upwards [hlo', R.eventually_b_le_r] with n hn hn'
+  intro i
+  simp only [Real.norm_of_nonneg (by positivity : 0 ≤ (n : ℝ))] at hn
+  calc b R.min_bi / 2 * n = b R.min_bi * n - b R.min_bi / 2 * n := by ring
+                          _ ≤ b R.min_bi * n - ‖n / log n ^ 2‖  := by gcongr
+                          _ ≤ b i * n - ‖n / log n ^ 2‖         := by gcongr; aesop
+                          _ = b i * n - n / log n ^ 2 := by
+                                congr
+                                exact Real.norm_of_nonneg <| by positivity
+                          _ ≤ r i n := by exact hn' i
```

Comment:
Unfortunately this is not true; it works in this particular example, but removing it in calc blocks often doesn't work (for example around line 253). In fact I started using `by exact` more or less systematically in calc blocks to avoid this bug. Maybe I should bring it up on Zulip...